### PR TITLE
fix: Throttle upload/download progress events to 400ms

### LINKED ISSUE #529: Debounce or rate-limit progress events on p2p upgrade downloads
I think some logic that limits the code to only emit on each whole percent (%) integer. The code for this is on this line: https://github.com/digidem/mapeo-mobile/blob/share-apk-p2p/src/backend/lib/upgrade-server.js#L235

cc @gmaclennan -- is this something you could pick up?

### PR DESCRIPTION
**Pull request by gmaclennan**
Fixes #529 using the [`progress-stream`](https://www.npmjs.com/package/progress-stream) module for progress events to only emit progress every 100ms.

### CODE REVIEW

**Comment by hackergrrl** [comment]
On 05/10  13:47, Gregor MacLennan wrote:
> Fixes #529 using the [`progress-stream`](https://www.npmjs.com/package/progress-stream) module for progress events to only emit progress every 100ms.

Nice work!

Nitpicky, but: 400-500ms is probably more than enough for it to look
smooth /wo wasting re-render cycles.

**Comment by gmaclennan** [comment]
> Nice work! Nitpicky, but: 400-500ms is probably more than enough for it to look smooth /wo wasting re-render cycles.

ok, done, and rebased. Please go ahead and merge if you think this is good to go.

File: src/backend/lib/upgrade-download.js [hunk 1]
```diff
@@ -5,14 +5,17 @@ const EventEmitter = require("events").EventEmitter;
...
 const searchLog = require("debug")("p2p-upgrades:search");
 const downloadLog = require("debug")("p2p-upgrades:download");
 const checkLog = require("debug")("p2p-upgrades:check");
 const clone = require("clone");
+const progressStream = require("progress-stream");
 
 const { DISCOVERY_KEY, UpgradeState } = require("./constants");
 
+// How frequently to emit progress events (in ms)
+const PROGRESS_THROTTLE = 400; // milliseconds
```
Comment (hackergrrl) [inline src/backend/lib/upgrade-download.js]:
```suggestion
const PROGRESS_THROTTLE_MS = 400;
```

File: src/backend/lib/upgrade-download.js [hunk 2]
```diff
@@ -225,14 +228,12 @@ class Download extends EventEmitter {
...
-        const progress = through((chunk, enc, next) => {
-          sofar += chunk.length;
-          this.setState(UpgradeState.Download.Downloading, {
-            sofar,
-            total: option.size,
-          });
-          next(null, chunk);
+        const progress = progressStream({
+          length: option.size,
+          time: PROGRESS_THROTTLE, // ms between each progress event
```
Comment (hackergrrl) [inline src/backend/lib/upgrade-download.js]:
```suggestion
          time: PROGRESS_THROTTLE_MS, // ms between each progress event
```

Just saves us from having to specify 'ms' in every usage in a comment.

File: src/backend/lib/upgrade-server.js [hunk 3]
```diff
@@ -16,6 +16,9 @@ type UploadInfo = {
 }
 */
 
+// How frequently to emit progress events (in ms)
+const PROGRESS_THROTTLE = 400; // milliseconds
```
Comment (hackergrrl) [inline src/backend/lib/upgrade-server.js]:
```suggestion
const PROGRESS_THROTTLE_MS = 400; // milliseconds
```

File: src/backend/lib/upgrade-server.js [hunk 4]
```diff
@@ -230,10 +232,13 @@ class UpgradeServer extends EventEmitter {
         this.uploads.push(upload);
         this.setState(this.state, this.uploads);
 
-        const tracker = through((chunk, enc, next) => {
-          if (chunk) upload.sofar += chunk.length;
+        const tracker = progressStream({
+          length: option.size,
+          time: PROGRESS_THROTTLE, // ms between each progress event
```
Comment (hackergrrl) [inline src/backend/lib/upgrade-server.js]:
```suggestion
          time: PROGRESS_THROTTLE_MS, // ms between each progress event
```

**Review by hackergrrl — COMMENTED** [review]
Looks good!

**Pull request merged**